_posts/2020-02-09-single-responsibility-principle.md

```diff
@@ -0,0 +1,271 @@
+
+However, despite of the advantages of finer-grained monitoring and scaling, **splitting responsibilities between smaller services generally increases the burden of operating the system.** Smaller services mean more work:
+- Setting up and operating intermediate message queues (like Kafka) between the services.
+ - DevOps: setting up and managing separate delivery pipelines, monitoring, configuration, machine and container images.
+ - Deployment and orchestration: Kubernetes doesn't fully alleviate it.
+ - To ensure [rollback safety](https://aws.amazon.com/builders-library/ensuring-rollback-safety-during-deployments/), the deployments should be multi-phase, shared state and messages sent between services should be versioned.
+
+### Reliability
+Reliability is the first software quality in the list that we mostly hurt, not aid when we split smaller responsibilities between the components. Although, if engineered properly, microservices can imporve availability: when one service is sick, others might still serve something for the users, the inherent [fallability of distributed systems](https://en.wikipedia.org/wiki/Fallacies_of_distributed_computing) hits harder. Also, "if engineered properly" is an important caveat :)
+
+Discussing the pros and cons of microservices is not the main goal of this article, but there are plenty of good materials on this topic, the reliability aspect in particular: [1](https://www.martinfowler.com/articles/microservice-trade-offs.html), [2](https://dwmkerr.com/the-death-of-microservice-madness-in-2018/), [3](https://developer.ibm.com/technologies/microservices/articles/challenges-and-benefits-of-the-microservice-architectural-style-part-1/).
+
+### Code Size
+Smaller responsibility of each component means that there are more components in total in the system.
+
+Each method needs a signature declaration. Each class needs constructors, static factory methods, field declarations, and other ceremony. Each module needs a separate configuration class and a depedendency injection setup. Each service needs separate configuration files, startup scripts, CI/CD/orchestration infrastructure, and so on.
+
+Therefore, **the more focused responsibilities of the components we make, the more code we will need to write.** This impacts the long-term maintainability much less than all the qualities discussed above: undertandability, flexibility, reusability, etc. However, it means that **it takes more time and effort to develop the first version of the application with finely separated responsibilites than with larger components**, especially if thorough testing is not the first priority as well.
```

Comment:
Yes! Building clean software takes longer initially but will save time later! I said similar things in my book :)